### PR TITLE
Implement tracing support with associated type on LogLevel trait

### PR DESCRIPTION
**Pull request by joshka**
Rationale: https://github.com/clap-rs/clap-verbosity-flag/issues/121#issuecomment-2475026188

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -189,148 +154,211 @@ pub trait LogLevel {
...
-            );
-            assert_eq!(
-                v.log_level_filter(),
-                *expected_filter,
-                "verbose = {verbose}, quiet = {quiet}"
-            );
-        }
-    }
-}
+// #[cfg(test)]
```
Comment (joshka) [inline src/lib.rs:157]: commented for now while wip

File: src/log.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+use crate::{LogLevel, Verbosity, VerbosityFilter};
+
+impl<L: LogLevel<LevelFilter = LevelFilter>> Verbosity<L> {
+    /// Get the log level.
+    ///
+    /// `None` means all output is disabled.
+    pub fn log_level(&self) -> Option<Level> {
+        self.log_level_filter().to_level()
+    }
```
Comment (joshka) [inline src/log.rs:12]: To move this up into the shared implementation (and the equivalent `tracing_level` method below), requires that we can make a generic type bound that converts the associated LevelFilter type to an associated Option<Level> type for both tracing and log crates. The log crate does not (yet) implement the necessary conversion methods. I've submitted a PR [to the log crate that should address that problem](https://github.com/rust-lang/log/pull/644) however. If that gets merged, then there is no type specific code differences on Verbosity left, and the entire implementation can work by just leveraging the conversion methods.

I think it would be reasonable to wait and see where that PR goes before completing this PR, as that would be a fairly neat outcome I think.

**Comment by epage** [comment]
What seems logical to me for commits
1. Switch to a new type for `VerbosityFilter` (this is a pre-requisite for `log` being optional, no associated types used)
2. Make `log` optional
3. Add optional `tracing` support

**Comment by joshka** [comment]
Sounds reasonable. I started with the premise that it wasn't necessary as a wrapper on u8, but as an enum instead that maps onto LevelFilter it's pretty neat.

Ooh, idea. If instead of trying to convert from LevelFilter to Option<Level> I just add conversions from VerbosityFilter, then this is actually possible to acheive without that log PR.

**Comment by epage** [comment]
> I started with the premise that it wasn't necessary as a wrapper on u8, but as an enum instead that maps onto LevelFilter it's pretty neat.

I'm fine with it being an `enum` instead of a `u8`.

> Ooh, idea. If instead of trying to convert from LevelFilter to Option I just add conversions from VerbosityFilter, then this is actually possible to acheive without that log PR.

Not quite sure what you are referring to with this idea

**Comment by joshka** [comment]
> Not quite sure what you are referring to with this idea

I wanted to move https://github.com/clap-rs/clap-verbosity-flag/blob/dddcca03bd419e50144be38356b74e42eb597d98/src/log.rs#L6-L13 into the main file rather than being log specific by converting `log::LevelFilter` to `Option<log::Level>` and `tracing_core::LevelFilter` to `Option<tracing_core::Level>`, the latter implements a conversion trait, the former doesn't. The log PR I mentioned addresses that, but it's not actually necessary. Instead I can add a type bound that converts from `VerbosityFilter` to `Option<L::Level>`, and implement that for the two crates. This would mean that the code in log.rs/tracing.rs is just conversion methods and log level implementations.

**Comment by joshka** [comment]
I re-implemented this with smaller clean logical commits and a few changes to naming etc.

**Comment by joshka** [comment]
I don't grok why this error is showing up. If the default features are disabled, then the log and tracing modules should not be imported, and hence should not have any doc tests run. I'm guessing there's some weird edge-case for this that I'm missing the rationale for right now. Any ideas?

```
running `cargo test --no-default-features` on clap-verbosity-flag (1/6)
     Compiling clap-verbosity-flag v2.2.2 (/home/runner/work/clap-verbosity-flag/clap-verbosity-flag)
      Finished `test` profile [unoptimized + debuginfo] target(s) in 0.22s
       Running unittests src/lib.rs (target/debug/deps/clap_verbosity_flag-aa91d2ce2ac824dc)
  
  running 0 tests
  
  test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
     Doc-tests clap_verbosity_flag
  error[E0432]: unresolved import `log`
   --> src/log.rs:4:9
    |
  4 | pub use log::Level;
    |         ^^^ help: a similar path exists: `crate::log`
    |
    = note: `use` statements changed in Rust 2018; read more at <https://doc.rust-lang.org/edition-guide/rust-2018/module-system/path-clarity.html>
  
  error[E0432]: unresolved import `log`
   --> src/log.rs:5:9
    |
  5 | pub use log::LevelFilter;
    |         ^^^ help: a similar path exists: `crate::log`
    |
    = note: `use` statements changed in Rust 2018; read more at <https://doc.rust-lang.org/edition-guide/rust-2018/module-system/path-clarity.html>
  
  error[E0432]: unresolved import `tracing_core`
   --> src/tracing.rs:4:9
    |
  4 | pub use tracing_core::{Level, LevelFilter};
    |         ^^^^^^^^^^^^ use of undeclared crate or module `tracing_core`
```

**Comment by joshka** [comment]
ACK on the doc tests. Will take a look at these tomorrow / in a couple of days.

File: src/lib.rs [hunk 3]
```diff
@@ -111,55 +111,143 @@ impl<L: LogLevel> Verbosity<L> {
...
+            1 => Filter::Error,
+            2 => Filter::Warn,
+            3 => Filter::Info,
+            4 => Filter::Debug,
+            5.. => Filter::Trace,
+        }
+    }
+}
+
+impl fmt::Display for Filter {
```
Comment (epage) [inline src/lib.rs:223]: Why do we have a `Display` impl?

**Comment by epage** [comment]
If we're stacking PRs, can we make sure that the lower PR (#123) is getting the latest updates so I can focus the review for that part there?

File: src/lib.rs [hunk 4]
```diff
@@ -191,70 +230,24 @@ impl fmt::Display for Filter {
...
+{
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         write!(f, "{}", self.filter())
     }
 }
 
 /// Customize the default log-level and associated help
 pub trait LogLevel {
+    type Level;
+    type LevelFilter;
```
Comment (epage) [inline src/lib.rs:250]: > no associated types used

With `Filter`, we shouldn't need any of this

Comment (joshka) [reply src/lib.rs]: It was either use associated types, or create crate specific methods (e.g. `log_level`, `tracing_level`). I think the specific methods are a much simpler approach.

7b8b965

**Comment by joshka** [comment]
This PR subsumes #123 as it's simpler to just do that with the macroed impl.

I'm guessing that you probably would like the commits reordered to go straight to the multiple methods approach here?

**Comment by epage** [comment]
> This PR subsumes https://github.com/clap-rs/clap-verbosity-flag/pull/123 as it's simpler to just do that with the macroed impl.

Conditionally present trait methods on a non-sealed trait is an improper use of features because a feature can be enabled in another crate while I implement only the part of the trait I use.  It could be argued that this crate isn't expected to be used in that way because this is for end-applications.  However, a workspace could have a mixture of applications and a `cargo check --workspace` will unify them.

I had been under the assumption that the trait would return our custom filter type, making it independent of any of the logging interfaces, and not needing macros.

I would also prefer to handle these concerns in separate APis rather than merging them.

File: src/lib.rs [hunk 5]
```diff
@@ -110,63 +124,107 @@ impl<L: LogLevel> Verbosity<L> {
...
-        Some(Level::Warn) => 2,
-        Some(Level::Info) => 3,
-        Some(Level::Debug) => 4,
-        Some(Level::Trace) => 5,
-    }
+/// A representation of the log level filter.
+///
+/// Used to calculate the log level and filter.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Filter {
```
Comment (joshka) [inline src/lib.rs:139]: Do you think this should be named `LevelFilter`?

I avoided doing so as this crate previously re-exported `log::LevelFilter` at the root, so this might be a bit of a confusing change.

File: src/lib.rs [hunk 6]
```diff
@@ -110,63 +124,107 @@ impl<L: LogLevel> Verbosity<L> {
...
+            Filter::Trace => write!(f, "trace"),
+        }
     }
 }
 
 /// Customize the default log-level and associated help
 pub trait LogLevel {
     /// Base-line level before applying `--verbose` and `--quiet`
-    fn default() -> Option<Level>;
+    fn default() -> Filter;
```
Comment (joshka) [inline src/lib.rs:199]: Would it make sense to rename this to `filter()` to avoid confusion with `Default::default()`, and app implementations of `LogLevel` that might wish to implement both traits for their apps?

File: src/lib.rs [hunk 7]
```diff
@@ -189,148 +247,56 @@ pub trait LogLevel {
...
-/// Default to [`log::Level::Error`]
-#[allow(clippy::exhaustive_structs)]
-#[derive(Copy, Clone, Debug, Default)]
-pub struct ErrorLevel;
-
-impl LogLevel for ErrorLevel {
-    fn default() -> Option<Level> {
-        Some(Level::Error)
-    }
+macro_rules! def_log_levels {
```
Comment (joshka) [inline src/lib.rs:222]: This saves about 40 lines of repetitive code. It's not strictly necessary, but I think that's a reasonable thing.

**Comment by epage** [comment]
> Can you clarify this - I'm not sure if by "separate APIs" you're agreeing with what I wrote above about having methods on Verbosity for each crate, or if you mean something else here.

Sorry, I meant "separate PRs".

**Comment by epage** [comment]
Also, please cleanup the commit history for how you'd like it reviewed and merged.  I know each maintainer on this is different but I find it difficult to review a PR when the commits go in different directions.

**Comment by joshka** [comment]
> Also, please cleanup the commit history for how you'd like it reviewed and merged. I know each maintainer on this is different but I find it difficult to review a PR when the commits go in different directions.

Yep - will do. Are you happy for me to do that once we've worked out the exact direction this ends up in?

**Comment by epage** [comment]
> Yep - will do. Are you happy for me to do that once we've worked out the exact direction this ends up in?

The other PR is pretty close to being merged, let's finish that up.

File: src/lib.rs (hunk 5 above)
Comment (epage) [reply src/lib.rs:139]: Good point.  This crate is focused around "verbosity" so your previous idea of having `Verbosity` in the name works.  I suspect we should be consistent with the other crates in calling this a `Filter` but a part of me finds that weird and would prefer `Level` especially since we only have "what level to show" and not "what level is this record" and have to deal with distinguishing them.

File: src/lib.rs (hunk 6 above)
Comment (epage) [reply src/lib.rs:199]: Good point about `Default::default`.  However, the name should convey that this is the starting point.  We could call it `default_filter`.

File: src/lib.rs (hunk 7 above)
Comment (epage) [reply src/lib.rs:222]: My bar for the use of Macros is high and I don't think this meets it.

File: src/lib.rs (hunk 5 above)
Comment (joshka) [reply src/lib.rs:139]: The filter part of the name communicates that `Off` is valid. Where-as there is no `Off` level that logs / trace events can be written at. I think Filter definitely belongs in the name because of this, it's mostly whether a prefix makes sense. I moved away from VerbosityFilter as it seemed like unnecessary repetition, but I think you might be right that it helps distinguish the filter as being distinctly this crate's Filter type.

**Comment by joshka** [comment]
Split into 3 stacked PRs:
- #125
- #126
- #127

File: src/lib.rs (hunk 7 above)
Comment (joshka) [reply src/lib.rs:222]: My threshold for this is a bit lower than yours. IMO, if a macro save a reasonable amount of code that I'll have to read / maintain without adding too much complexity then it's worth including. This is because they make things that look similar actually be similar. Usually that's somewhere around 5 repetitions of 3 things (e.g. struct definition, associated derives, implementing a trait hits my threshold pretty niicely).

I'm happy to go with your threshold here, though I'd suggest giving some future thought about whether that could do with a slight adjustment. If not, no big deal :)

File: src/lib.rs (hunk 5 above)
Comment (joshka) [reply src/lib.rs:139]: in #125 this became VerbosityFilter

File: src/lib.rs (hunk 6 above)
Comment (joshka) [reply src/lib.rs:199]: Changed in #125
